### DISCUSSION TITLE
Reverse proxy requires Host to be set in headers

### DISCUSSION THREAD

**Discussion by grisu48** [LycheeOrg/Lychee#2156] [Q&A] [answered]
Hey people.

Since some time my https://photo.feldspaten.org instance is broken, probably due to the v5 upgrade.

I run lychee as a `podman` container with `docker.io/lycheeorg/lychee:latest`. I have the `APP_URL` configured and pointing to `https://photo.feldspaten.org`, both as environment variable for `podman` and present in the `/conf/.env` file within the container.

However, if I check the resource URLs on https://photo.feldspaten.org then they all point to the container ip `10.88.0.12` and I don't know what I'm doing wrong here.

I check the env variables within the container, and `APP_URL` is set properly: 

```
# podman exec -ti lychee /bin/env
container=podman
USER=lychee
PHP_TZ=Europe/Amsterdam
PUID=1000
PGID=1000
APP_URL=https://photo.feldspaten.org
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=photo.feldspaten.org
TERM=xterm
HOME=/root
```

The release notes state explicitly that starting from v5 the `APP_URL` needs to be set, which I did, so I fail to see why this is still broken. Anyone has an idea?

**Comment by ildyria** [comment]
You are mistaking `.env` and your environment variables.
There should be a `.env` file at the root of Lychee with those. That is also where you set up your database connection.

Because you are using LycheeOrg/Lychee, you should have a `lychee/conf` folder in which you will find your `.env` that is where you want to apply your changes.

**Reply by grisu48** [reply]
Thanks for the answer. I did apply `APP_URL` both, as environmental variable and in `.env` as you can see here:

```
# podman exec -ti lychee /bin/bash
root@photo:/var/www/html/Lychee# grep 'APP_URL' .env
APP_URL=https://photo.feldspaten.org
root@photo:/var/www/html/Lychee# 
```

For completeness, I start the container as systemd service as follows:

```
ExecStart=/usr/bin/podman container run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --sdnotify=conmon --replace \
        --name lychee \
        --hostname photo.feldspaten.org \
        --env PHP_TZ=Europe/Amsterdam \
        --env APP_URL=https://photo.feldspaten.org \
        --volume /mnt/RAID/lychee/conf:/conf:Z \
        --volume /mnt/RAID/lychee/uploads:/uploads:Z \
        --volume /mnt/RAID/lychee/sym:/sym:Z \
        --ip 10.88.0.12 \
        --label io.containers.autoupdate=image \
        --detach=True \
        --memory 2g \
        docker.io/lycheeorg/lychee:latest
```

I have a `.env` file in `/mnt/RAID/lychee/conf`, which ends up in `/conf/.env` in the container.
And in the container there is a symlink `/var/www/html/Lychee/.env -> /conf/.env`. The contents of the symlink file match the contents of the `.env` file in my `/conf` directory.

From what I can see the `APP_URL` is in the `.env` file (it always was) but still it won't work - what could I have done wrong here?

**Reply by ildyria** [reply]
Reverse proxy messing up maybe ? I find that one very surprising.

**Reply by grisu48** [reply, marked as answer]
Yes, that was it! I needed to set the `Host` header in my nginx configuration:

```nginx
location / {
    proxy_pass http://10.88.0.12/;   # Container IP
    proxy_set_header Host            $host;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Proto https;
    client_max_body_size 128M;
}
```

The important setting is the `Host` setting, the `X-Forwarded-For` is not strictly required to work. So in addition to the `APP_URL` also the `Host` HTTP header needs to be set correctly. This was not an issue in v4.

I also needed to set `APP_FORCE_HTTPS=true` in the `.env` settings but that's a different issue and might depend on your configuration.

EDIT: I added `proxy_set_header X-Forwarded-Proto https;` following https://github.com/LycheeOrg/Lychee/discussions/2156#discussioncomment-8023421

**Reply by grisu48** [reply]
@ildyria I changed the title and marked the answer accordingly. Feel free to edit and/or comment to further tidy this correspondence.

And thanks for your work! :balloon:

**Reply by ildyria** [reply]
Can you try to also add 
```
        proxy_set_header X-Forwarded-Proto https;
```

and remove `APP_FORCE_HTTPS=true` that should be equivalent and cleaner. :)

**Reply by grisu48** [reply]
Yep, that appears also to work. Nice!

**Reply by grisu48** [reply]
The only thing is that it doesn't work well with paths. I have another instance, e.g. `https://hostname.local/lychee` and that forwards to `https://hostname.local/gallery` instead of `https://hostname.local/lychee/gallery`. I'm looking on how to fix that ones as well now.

**Reply by grisu48** [reply]
Ah, https://github.com/LycheeOrg/Lychee/pull/2150 addresses this issue. Will wait for a new release then. Thanks!

**Reply by ildyria** [reply]
> The only thing is that it doesn't work well with paths. I have another instance, e.g. `https://hostname.local/lychee` and that forwards to `https://hostname.local/gallery` instead of `https://hostname.local/lychee/gallery`. I'm looking on how to fix that ones as well now.

IMHO, a better approach would be `https://gallery.hostname.local` or something like that :)

**Reply by grisu48** [reply]
True, but that's not a possible option in this case (can't control subdomains for that hostname).

**Reply by Szparki** [reply]
I was wondering if you can share your nginx configuration for Lychee on docker behind nginx? I've been working on it for awhile now and I'm still getting upload failed and error 401 in nginx log.
<img width="482" alt="lychee error" src="https://github.com/user-attachments/assets/d5196eb8-89fc-43b4-b84f-f5117fe2f1dd">

**Reply by grisu48** [reply]
It's basically https://github.com/LycheeOrg/Lychee/discussions/2156#discussioncomment-8023398 with some timeout values that likely don't make a big difference

```nginx

server {
	# ... listen, and ssl stuff ...

	location / {
		proxy_pass http://lychee.podman/;
		proxy_set_header Host            $host;
		proxy_set_header X-Forwarded-For $remote_addr;
		proxy_set_header X-Forwarded-Proto https;

		client_max_body_size        256M;
		proxy_connect_timeout       30;
		proxy_send_timeout          300;
		proxy_read_timeout          300;
		send_timeout                300;
	}
}
```

I have `lychee.podman` in my `/etc/hosts`, you can use an IP address there as well.

**Reply by Szparki** [reply]
Thank you @grisu48, I tried this and I'm still getting 401 when I'm trying to upload a file
```
[12/Jul/2024:11:56:23 +0000] "POST /livewire/upload-file?expires=1720788982&signature=c141333d9e1b3efee6cefd4063373d1b9443dffe69ab661aa8108e8c46e4f32c HTTP/1.1" 401 19181 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:127.0) Gecko/20100101 Firefox/127.0"
```
Can you share please your docker compose file too?

**Reply by grisu48** [reply]
I run it as a `podman` container using [quadlet](https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html#template-files). Not sure if this is helpful for you, but here it goes:

```ini
[Unit]
Description=lychee
Wants=network-online.target
After=network-online.target
RequiresMountsFor=%t/containers
StartLimitBurst=2
StartLimitIntervalSec=120

[Container]
Image=docker.io/lycheeorg/lychee:v5.5.1
ContainerName=lychee
AutoUpdate=registry
IP=10.88.0.10

Volume=/srv/lychee/conf:/conf:Z
Volume=/srv/lychee/uploads:/uploads
Volume=/srv/lychee/sym:/sym

Environment=PHP_TZ=Europe/Amsterdam
Environment=ENABLE_TOKEN_AUTH=false

#HealthCmd=curl --head --fail http://10.88.0.10/
#HealthStartPeriod=2m

ShmSize=256m
PodmanArgs=--memory 2g

[Service]
Restart=always

[Install]
WantedBy=multi-user.target default.target
```
